### ISSUE TITLE
HAR Schema Questions & Discussion

### ISSUE THREAD

**Issue by thibaultcha** [Kong/httpsnippet#9]
I have several questions after implementing objc native and unirest (coming):
- **What handles or should handle schema errors?** It seems weird having to make checks such as:

``` js
if (this.source.postData && (this.source.postData.params || this.source.postData.text)) {
```

Right now it is possible to have a `postData` without `params` or `text`. This should probably throw an error to the user saying his HAR description is wrong? For this reason (and because the fixtures for unit tests probably don't cover enough cases) Node native is currently wrong, see #8. And it makes us write annoying checks instead of a simple... say `switch` for form-encoded and others.
- **Cannot build a native JSON object when mimeType is 'application/json'** because the `text` property is a string. Having to send JSON as a string where the majority of other examples on the Internet are object literals seems... weird. It also serves well enough its purpose for a copy-pasted code snippet but the programmer cannot add logic on top of it, or intuitively add a parameter.

Example for obj-c sending json like most people do:

``` objective-c
// ...
NSDictionary *parameters = @{@"parameter": @"value", @"foo": @"bar"};
NSData *postData = [NSJSONSerialization dataWithJSONObject:parameters options:0 error:&error];
[request setHTTPBody:postData];
```

But httpsnippet makes us send stringified JSON manually encoded:

``` objective-c
NSData *postData = [[NSData alloc] initWithData:[@"{\"foo\": \"bar\"}" dataUsingEncoding:NSUTF8StringEncoding]];
[request setHTTPBody:postData];
```
- **No way to know the response type in advance for sure?** For Unirest the class of the response object will vary, because one will deserialize the response if JSON (`UNIHTTPJsonResponse`) and the other will return it as is (`UNIHTTPStringResponse`) (same for .NET I think). Sure I can rely on the `Accept` header but who says it will be acknowledged?

it's not so much about the class of the response because when you think about it, we are missing opportunities to show how to handle JSON responses in every other language (where JSON is the most used response type)

**Comment by ahmadnassri** [comment]
> What handles schema errors?
1. [har-validator](https://www.npmjs.com/package/har-validator) is used to validate the HAR input based on [HAR 1.2 Spec](http://www.softwareishard.com/blog/har-12-spec/)
2. in [`src/index.js`](https://github.com/ahmadnassri/httpsnippet/blob/master/src/index.js#L16-L25) I specifically add any missing input properties _before_ validation, to make it simpler and faster to create input json files _(rather than having to add in an empty `cookies: []` array, and specificy `httpVersion` every time for example)_

> It seems weird having to make checks such as:
> 
> ``` js
> if (this.source.postData && (this.source.postData.params || this.source.postData.text)) {
> ```

yes, it is, and you shouldn't be doing that, at least, it seems you're approaching the issue thinking only of Unirest. there are libraries that allow you to set a raw body **AND** set parameterized POST data at the same time ... I see no reason for httpsnippet to limit that behavior ... it should be left to the discretion of the implementer to use what makes sense for the library.

keeping the implementation simple, it would be enough to check that `.text` and `.params` are mutually exclusive:

``` js
if (this.source.postData.text) {
  // request body
}

if (!this.source.postData.text && this.source.postData.params) {
  // construct post params
}
```

> Cannot build a native JSON object when mimeType is 'application/json'

other than JavaScript, I'm not aware of any language that has **native** JSON support, so therefore if you want to implement whatever **native equivalent** of JSON, you probably have to parse it first with `JSON.prase`, then process / convert it to that native as well ... I don't see any way around that?

> No way to know the response type in advance? 

I'm going to be soon adding the ability to reference a FULL HAR object / file rather than just the request portion of it, which should give the ability to review the [response object](http://www.softwareishard.com/blog/har-12-spec/#response) and construct accordingly.

**Comment by thibaultcha** [comment]
> yes, it is, and you shouldn't be doing that

No, I have no other choice. It makes me do even uglier things for Unirest. I need a way to know wheter or not a request as a body. And it seems that the existence of `postData` actually doesn't mean anything.

> I'm not aware of any language that has native JSON support

I should have put native in quotes but you got the idea.

> so therefore if you want to implement whatever native equivalent of JSON, 

I know I could parse it, I'm just saying now I have to do it for every language because your schema doesn't already do it...

Basically I know your points. I am just trying to start a discussion about whether or not HAR is the right schema to handle snippets creation honestly... I find the Mashape one way more appropriate.

**Comment by thibaultcha** [comment]
> har-validator is used to validate the HAR input based on HAR 1.2 Spec

I know that. Same thing, I'm just saying we should add a layer of error handling. When the HAR format is valid but the informations are not enough to properly build a request snippet. (see the postData example, and #8)

**Comment by ahmadnassri** [comment]
>  And it seems that the existence of postData actually doesn't mean anything.

yes, as mentioned, `postData` is always generated regardless of source, I still think you're not looking closer at the HAR spec, here's another valid way of doing body work:

``` js
if (this.source.postData.text) {
  // request body
} else {
  if (this.source.postData.params) {
    // construct post params
  }
}
```

basically the existence of `postData.text` overrides everything _(and is the easiest, just sent `content-type` header and send the string of text)_

also, `postData.params` are not JUST the key/value pair for `application/x-www-form-urlencoded` ... in the case of `multipart/form-data` you have to traverse each param, and parse it's value individually as some would be a file, others can be any content (text, video, etc ...)

ex:

``` js
"params": [
    {
        "name": "file",
        "value": "[content of the pdf file]",
        "fileName": "example.pdf",
        "contentType": "application/pdf",
    }
]
```

in this case you should parse the `fileName` field and use that in the code (to look up and read content of the file) then pass that content as the `multipart/form-data` request with the specified name.

> I know I could parse it, I'm just saying now I have to do it for every language because your schema doesn't already do it...

I could pre-parse the JSON object into a property called `this.source.json` or something ... but then I would also have to traverse all the potential `multipart/form-data` params and look for any existence of `application/json` (or any of the 5 other headers representing JSON) and parse those as well ... its doable... but seems like a lot ... any better suggestions?

> I find the Mashape one way more appropriate.

the Mashape schema is lacking in many places, it only covers the most popular use-cases and certainly does not deal with `multipart/form-data` at all!

**Comment by ahmadnassri** [comment]
this is a good discussion regardless, perhaps if any valuable lessons come out of this, we can share with the HAR community :)

**Comment by thibaultcha** [comment]
(Your code snippet for body work still does not do the job if one wants a definitive way of knowing whether or not a request has a body. Doing it like that, I would have to duplicate my code. The if statement mentioned in the first comment is still the only way of knowing whether or not the request has a body.)

True for the multipart requests, however I still believe the Mashape snippets are simpler (and they are actually way shorter (thanks also to the helpers)) but also because the schema is simply less flexible.

**Comment by ahmadnassri** [comment]
with recent guideline added and utility properties, this solves 90% if not all the of concerns discussed here...

**Comment by thibaultcha** [comment]
I approve this message.
